File: pdl/infra/php/src/Io/Pdl/Infrastructure/Data/RowExecutor.php
<?php

namespace Io\Pdl\Infrastructure\Data;

final class RowExecutor
{
    public static function create(Row $record): void
    {
        $values = $record->collectValues();
        $store = $record->resolveStore();
        $record->setStore($store);
        $inserted = $store->insert($record->table(), $record->primaryKey(), $values);
        if (!empty($inserted)) {
            $record->applyValues($inserted);
        }
    }

    public static function update(Row $record): void
    {
        $values = $record->collectValues();
        $store = $record->resolveStore();
        $record->setStore($store);
        $store->update($record->table(), $record->primaryKey(), $values);
    }

    public static function delete(Row $record): void
    {
        $values = $record->collectValues();
        $store = $record->resolveStore();
        $record->setStore($store);
        $store->delete($record->table(), $record->primaryKey(), $values);
    }

    public static function hydrate(Row $record, array $values): void
    {
        $record->applyValues($values);
    }
}

File: pdl/infra/kotlin/src/main/kotlin/io/pdl/infrastructure/data/RowExecutor.kt
package io.pdl.infrastructure.data

object RowExecutor {
    @JvmStatic
    fun create(record: Row) {
        val store = record.resolveStore()
        record.setStore(store)
        val values = record.collectValues()
        val inserted = store.insert(record.table(), record.primaryKey(), values)
        if (inserted.isNotEmpty()) {
            record.applyValues(inserted)
        }
    }

    @JvmStatic
    fun update(record: Row) {
        val store = record.resolveStore()
        record.setStore(store)
        val values = record.collectValues()
        store.update(record.table(), record.primaryKey(), values)
    }

    @JvmStatic
    fun delete(record: Row) {
        val store = record.resolveStore()
        record.setStore(store)
        val values = record.collectValues()
        store.delete(record.table(), record.primaryKey(), values)
    }

    @JvmStatic
    fun hydrate(record: Row, values: Map<String, Any?>) {
        record.applyValues(values)
    }
}
